### PR TITLE
* Added support for .move extents time range

### PR DESCRIPTION
**Pull request by asaharn**
#### Pull Request Description

_Improved performance by providing timerange to .move extents_

---

#### Future Release Comment
_[Add description of your change, to include in the next release]_
_[Delete any or all irrelevant sections, e.g. if your change does not warrant a release comment at all]_

**Breaking Changes:**
- None

**Features:**
- None

**Fixes:**
- None

### CODE REVIEW

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala [hunk 1]
```diff
@@ -128,6 +128,7 @@ object KustoWriter {
       val partitionsResults = rdd.sparkContext.collectionAccumulator[PartitionResult]
       val parameters = KustoWriteResource(authentication = authentication, coordinates = tableCoordinates,
         schema = data.schema, writeOptions = rebuiltOptions, tmpTableName = tmpTableName)
+      val sinkStartTime = Instant.now()
```
Comment (y0nil) [inline connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:131]: please note that in some clients/pipelines, customers can override the `creationTime` of extents created by ingestion as part of ingestion properties, in which case - the minimum and maximum values for the datetime range should be based on the properties they pass.
if this isn't supported in the Spark connector - then all is good.
if it is, then this will need to be adjusted accordingly (i.e. if `creationTime` is specified - use that instead of `Instants.now()`)

Comment (asaharn) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: Done.

Comment (y0nil) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: i think you misunderstood. let me try to explain differently
- you don't have to change the creationTime property. only in rare cases do customers choose to do so.
- if _the customer_ hasn't chosen to set the 'creationTime' property, then using the current datetime (UTC now) to capture the job's start/end time is just fine.

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala [hunk 2]
```diff
@@ -166,12 +167,20 @@ object KustoWriter {
...
         }
       }
     }
   }
 
+  def getCreationTime(ingestionProperties: SparkIngestionProperties, tableCoordinates: KustoCoordinates): Instant = {
+    val creationTime = Option(ingestionProperties.toIngestionProperties(tableCoordinates.database, tableCoordinates.table.get).getAdditionalProperties.get("creationTime"))
+    creationTime match {
+      case Some(creationTimeVal) => Instant.parse(creationTimeVal)
+      case None => Instant.now()
```
Comment (ag-ramachandran) [inline connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:180]: `Instant.now(Clock.systemUTC())` Is the best option to use to be sure.

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala [hunk 3]
```diff
@@ -166,12 +167,20 @@ object KustoWriter {
...
+          crp, tableExists, rdd.sparkContext, authentication, kustoClient, sinkStartTime)
         }
       }
     }
   }
 
+  def getCreationTime(ingestionProperties: SparkIngestionProperties, tableCoordinates: KustoCoordinates): Instant = {
+    val creationTime = Option(ingestionProperties.toIngestionProperties(tableCoordinates.database, tableCoordinates.table.get).getAdditionalProperties.get("creationTime"))
+    creationTime match {
+      case Some(creationTimeVal) => Instant.parse(creationTimeVal)
```
Comment (ag-ramachandran) [inline connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:179]: Do we validate this date time format upfront ? If we already have it in the documentation its okay. We have to explicitly call out and even better if we validate this upfront when we parse the options.
If not now, we can add a TODO that we need to add this at some point

**Review by ag-ramachandran — CHANGES_REQUESTED** [review]

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala (hunk 1 above)
Comment (ag-ramachandran) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: Please correct us here @y0nil . The logic was the following. If the user overrides the extentCreationTime, then we set it this way or else fallback to now. Did we miss something here.

```
  def getCreationTime(ingestionProperties: SparkIngestionProperties, tableCoordinates: KustoCoordinates): Instant = {
    val creationTime = Option(ingestionProperties.toIngestionProperties(tableCoordinates.database, tableCoordinates.table.get).getAdditionalProperties.get("creationTime"))
    creationTime match {
      case Some(creationTimeVal) => Instant.parse(creationTimeVal)
      case None => Instant.now()
```

Comment (y0nil) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: no, you don't need to explicitly set creationTime. only if the end-user specified is should it be set.
otherwise, it shouldn't be explicitly set.

Comment (asaharn) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: Yes. Agreed Yoni. I guess that's what we did here? We are not setting the creationTime anywhere, but just taking the creationTime (only if exists) in first place, else the current time.

Comment (y0nil) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: let me try to explain again - you should never set creationTime to "Instant.now()". you should only set it if it was provided by the end user

Comment (asaharn) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: I think you are confusing the local variable creationTime with the ingestion property `creationTime`. Sorry for the confusing name will change it. 
```
def getCreationTime(ingestionProperties: SparkIngestionProperties, tableCoordinates: KustoCoordinates): Instant = {
    val creationTime = Option(ingestionProperties.toIngestionProperties(tableCoordinates.database, tableCoordinates.table.get).getAdditionalProperties.get("creationTime"))
    creationTime match {
      case Some(creationTimeVal) => Instant.parse(creationTimeVal)
      case None => Instant.now()
```

The local variable, creationTime, will be set if the ingestion prop `creationTime` is provided by user. Else the current Instant will be set. and returned.

Comment (y0nil) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: yes, that's much clearer than the previous version you quoted

```
    creationTime match {
      case Some(creationTimeVal) => Instant.parse(creationTimeVal)
      case None => Instant.now()
```

Comment (ag-ramachandran) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala]: Thanks @y0nil, with the semantic change does it sound okay :) if so , will fold it into master and plan a release this week.

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/CslCommandsGenerator.scala [hunk 4]
```diff
@@ -81,22 +82,23 @@ private[kusto] object CslCommandsGenerator {
     s""".show table ${tableName} details | project todynamic(ShardingPolicy).UseShardEngine"""
   }
 
-  def generateTableMoveExtentsCommand(sourceTableName: String, destinationTableName: String, batchSize: Int,
+  def generateTableMoveExtentsCommand(sourceTableName: String, destinationTableName: String, timerange: Array[Instant], batchSize: Int,
                                       isDestinationTableMaterializedViewSource: Boolean = false): String = {
     val setNewIngestionTime: String = if (isDestinationTableMaterializedViewSource) "with(SetNewIngestionTime=true)" else ""
     s""".move extents to table $destinationTableName $setNewIngestionTime <|
-       .show table $sourceTableName extents with(extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":$batchSize}');
+       .show table $sourceTableName extents with(extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":$batchSize}') with(extentCreatedOnFrom='${timerange(0)}', extentCreatedOnTo='${timerange(1)}');
```
[the remark this answers is not in the document]
Comment (asaharn) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/CslCommandsGenerator.scala:89]: Hi Yoni,

Yes we have E2E tests that checks the functionality by writing the data and reading that back again to check if both are equal. And the above code passed them. 

After your suggestion above i.e. `.move extents to table $destinationTableName with(setNewIngestionTime=<true of false>, extentCreatedOnFrom='...', extentCreatedOnTo='...', extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":...}') <| .show table $sourceTableName extents`, 
got the below findings:

1) I took the final command built by code (in current scenario) and ran it manually which ran without any issues with the desired result. 
```kql
.move async extents to table KustoSparkReadWriteTest_f84ac15e_3f51_49cf_a6f0_bbb70915xxxx  <|
       .show table sparkTempTable_KustoSink_KustoSparkReadWriteTest_xxxxx_3f51_49cf_a6f0_bbb70915018f_3004f1b6_1599_49a8_91cc_xxxxx
       extents with(extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":1}')
       with(extentCreatedOnFrom='2023-03-02T11:28:41.474Z', extentCreatedOnTo='2023-03-02T11:29:02.405063200Z');
```

2) After making the changes asked above, i.e. moving all the props in a single with we got exceptions:
```kql
.move async extents to table KustoSparkReadWriteTest_f84ac15e_3f51_49cf_a6f0_bbb70915xxxx with ( extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":1}', extentCreatedOnFrom='2023-03-02T11:28:41.474Z', extentCreatedOnTo='2023-03-02T11:29:02.405063200Z') <|
       .show table sparkTempTable_KustoSink_KustoSparkReadWriteTest_xxxxx_3f51_49cf_a6f0_bbb70915018f_3004f1b6_1599_49a8_91cc_xxxxx
       extents
```
 got the below error:
image


Which make sense to me, since we cannot move prop `extentsShowFilteringRuntimePolicy` before `.show table extents` command. 

Really appreciate your suggestions, or to wrap this faster we can have a 15min call?

[the remark this answers is not in the document]
Comment (y0nil) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/CslCommandsGenerator.scala:89]: sorry for the confusion.

instead of this:

```
.move async extents to table KustoSparkReadWriteTest_f84ac15e_3f51_49cf_a6f0_bbb70915xxxx with ( extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":1}', extentCreatedOnFrom='2023-03-02T11:28:41.474Z', extentCreatedOnTo='2023-03-02T11:29:02.405063200Z') <|
       .show table sparkTempTable_KustoSink_KustoSparkReadWriteTest_xxxxx_3f51_49cf_a6f0_bbb70915018f_3004f1b6_1599_49a8_91cc_xxxxx
       extents
```

it should be this:

```
.move async extents to table KustoSparkReadWriteTest_f84ac15e_3f51_49cf_a6f0_bbb70915xxxx with (extentCreatedOnFrom='2023-03-02T11:28:41.474Z', extentCreatedOnTo='2023-03-02T11:29:02.405063200Z') <|
       .show table sparkTempTable_KustoSink_KustoSparkReadWriteTest_xxxxx_3f51_49cf_a6f0_bbb70915018f_3004f1b6_1599_49a8_91cc_xxxxx
       extents with(extentsShowFilteringRuntimePolicy='{"MaximumResultsCount":1}')
```

[the remark this answers is not in the document]
Comment (asaharn) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/CslCommandsGenerator.scala]: Thanks @y0nil. Have added the changes.

**Review by y0nil — APPROVED** [review]

**Review by ag-ramachandran — APPROVED** [review]
